### DISCUSSION TITLE
Repology times out in GH hosted app

### DISCUSSION THREAD

**Discussion by DucretJe** [renovatebot/renovate#25423] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

Hi,

2 days ago (Oct 23) renovate started to fail on one of my repos with failures on `repology`.

Here is an extract of the logs (no details) for my last run

<details><summary>Logs (no details)</summary>
<p>

```plain
INFO: Dependency extraction complete
DEBUG: RollingReleasesData - data written
DEBUG: Using queue: host=repology.org, concurrency=1
DEBUG: GET https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=bat = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60005)
WARN: Repology lookup failed with unexpected error
DEBUG: GET https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=build-essential-mipsen = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60040)
WARN: Repology lookup failed with unexpected error
DEBUG: GET https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=ca-certificates = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60059)
WARN: Repology lookup failed with unexpected error
DEBUG: GET https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=cmake = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60061)
WARN: Repology lookup failed with unexpected error
DEBUG: GET https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=containerd = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60039)
WARN: Repology lookup failed with unexpected error
DEBUG: GET https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=curl = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60042)
WARN: Repology lookup failed with unexpected error
DEBUG: GET https://repology.org/tools/project-by?repo=arch&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=docker = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60011)
WARN: Repology lookup failed with unexpected error
DEBUG: GET https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=dos2unix = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60048)
WARN: Repology lookup failed with unexpected error
DEBUG: GET https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=exa = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60027)
WARN: Repology lookup failed with unexpected error
```

</p>
</details> 

Here is a detailed extract for one of the failed dep:

<details><summary>Detailed extract</summary>
<p>

```plain
DEBUG: GET https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=cmake = (code=ETIMEDOUT, statusCode=-1 retryCount=2, duration=60061)
WARN: Repology lookup failed with unexpected error
{
  "packageName": "debian_12/cmake"
  "err": {
    "name": "TimeoutError",
    "code": "ETIMEDOUT",
    "timings": {
      "start": 1698223149763,
      "socket": 1698223149763,
      "lookup": 1698223149764,
      "error": 1698223209824,
      "phases": {
        "wait": 0,
        "dns": 1,
        "total": 60061
      }
    },
    "event": "request",
    "message": "Timeout awaiting 'request' for 60000ms",
    "stack": "RequestError: Timeout awaiting 'request' for 60000ms\n    at ClientRequest.<anonymous> (/opt/containerbase/tools/renovate/37.31.5/node_modules/got/dist/source/core/index.js:970:65)\n    at Object.onceWrapper (node:events:632:26)\n    at ClientRequest.emit (node:events:529:35)\n    at ClientRequest.emit (node:domain:489:12)\n    at ClientRequest.origin.emit (/opt/containerbase/tools/renovate/37.31.5/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (node:_http_client:501:9)\n    at TLSSocket.emit (node:events:517:28)\n    at TLSSocket.emit (node:domain:489:12)\n    at emitErrorNT (node:internal/streams/destroy:151:8)\n    at emitErrorCloseNT (node:internal/streams/destroy:116:3)\n    at processTicksAndRejections (node:internal/process/task_queues:82:21)\n    at Timeout.timeoutHandler [as _onTimeout] (/opt/containerbase/tools/renovate/37.31.5/node_modules/got/dist/source/core/utils/timed-out.js:36:25)\n    at listOnTimeout (node:internal/timers:571:11)\n    at processTimers (node:internal/timers:512:7)",
    "options": {
      "headers": {
        "user-agent": "RenovateBot/37.31.5 (https://github.com/renovatebot/renovate)",
        "accept": "application/json",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=cmake",
      "hostType": "repology",
      "username": "",
      "password": "",
      "method": "GET",
      "http2": false
    }
  }
}
```

</p>
</details> 

I tried to query the same endpoint:

```bash
curl -L "https://repology.org/tools/project-by?repo=debian_12&name_type=binname&target_page=api_v1_project&noautoresolve=on&name=cmake"                                                                                                                                                                                                   
```

No issue on my side.

One strange thing is that my dasboard doesn't shows any error.
I have the feeling that the full renovate job times out at some point. I have a lot of repology dependencies in this repository and it takes 60000ms for each one, I guess after a certain amount of time it just stops.

Reproduction repo:
https://github.com/DucretJe/renovate-minimal-reproduction

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
The repology owner has at times blocked IP addresses and this is the most likely answer

**Reply by rarkins** [reply]
@AMDmi3 Have you banned any AWS IP addresses recently?

**Reply by spalvarez** [reply]
@rarkins I'm having the same issue with https://github.com/ThoughtWorks-DPS/circleci-infra-gcp.  I'm able to query for updates from repology with no errors when running the renovate npm app from my local machine authorized to GitHub with a PAT token, but the GitHub app bot shows errors from timeouts on the developer.mend.io portal.

**Reply by AMDmi3** [reply]
Yes, I have banned some IPs recently for TOS violations.

**Reply by rarkins** [reply]
Each Renovate instance obeys the previously agreed approach. The challenge is that they all share one IP, so you may consider them to be one server violating your ToS.

This IP never changes, so maybe you can take another look? You can literally see our source code to confirm that we've done everything asked so far. I don't personally need Repology for my repos but every time you block Renovate it makes it painful for those who do.

**Reply by AMDmi3** [reply]
Why's there multiple instances? The reason for rate limitation was not a formality, but preventing excess site load, and if there are N instances with 1rps limit, it's effectively Nrps which does not comply with the ToS. I can see up to 7-8 requests in a single second, some to the same url, it's a waste of resources and causes user visible lags. If there are multiple instances, I have to ask to increase the delay proportionally.

**Reply by rarkins** [reply]
We run around 300 free Renovate instances for users, each on its own server. They share a cache, but:
- caches expire, need refreshing
- there's some chance two instances could see the same expired entry at the same time and both try, but that should be rare

Was it peak rps which was the main problem or sustained? i.e. it spiked at times and that was the problem? Or simply that overall the number of users we have these days is a problem for you in general?

I understand that it's a free service, but so too is Renovate. 100% of the repology traffic you see is by users opting into it through config, and not because we send traffic by default.

**Reply by AMDmi3** [reply]
The problem are moments when a batch of heavy requests stack up and take all processing time, and users experience visible lag of a few seconds. I see these myself lately, but I haven't measured these. It is likely that peaks are more of a problem. For the record 45% of renovate load from that AWS IP was 2 RPS, 32% was 3 rps and 2% was 4-8 rps. I guess that I can just set up stricter rate limit on repology side. Please remind me whether renovate handles 429s correctly.

**Reply by AMDmi3** [reply, marked as answer]
(IP unblocked in the meantime)

**Reply by DucretJe** [reply]
I confirm it's unblocked, thanks a lot

**Reply by rarkins** [reply]
I have also added this rule so that each host makes at most one request for every 2 seconds: https://github.com/renovatebot/renovate/commit/c39e67faf3d4b72dba5ce8220b78a99941049d43

It's possible that users can skip this config though, so we need to also think how to apply it to all

**Reply by HonkingGoose** [reply]
## Renovate should handle `429`'s in some way already

@AMDmi3 said:

> Please remind me whether renovate handles 429s correctly.

Renovate should already handle `429`'s in _some_ way, we apparently use the `got` library, which should support `retry-after` out-of-the-box. See the comments in this issue:

- #9550

## Do we have a "improve `429` methods" issue already?

`@rarkins` said in #9550:

> Seeing as were apparently already doing it.. [dealing with `429`'s] I'm going to consider this issue [#9550] closed. Separate issues welcome for changing the list of methods to retry, etc.

I can't find any open issues when searching for `429` or `retry-after`. Maybe one of the maintainers can find a related issue?

I remember mentioning the "exponential back-off algorithm" as a possible way to reduce load on a server. I don't know if `got` already does this, or if we need to change the way we use the library.

To me it seems like we need a new discussion for the `429` and/or `retry-after` stuff? I'll let the maintainers make the discussion. 😄 

### Should we link to a new/other issue in the Repology workaround `description`?

On a related note, the `description` for the Repology workaround says:

```
    description:
      'Limit requests to reduce load on Repology servers until we can fix this properly, see issue `#10133`.',
```

We "closed" issue #10133 with PR #10146 which added the workaround preset from above. Because we only added a workaround preset, it doesn't feel to me that we properly closed the issue. 😄 Maybe we should link to a new discussion or issue?

**Reply by rarkins** [reply]
Added #25499
